pkg/network/usm/ebpf_gotls.go

```diff
@@ -399,33 +430,38 @@ func (p *goTLSProgram) handleProcessStart(pid pid) {
 
 // addInspectionResultToMap runs a binary inspection and adds the result to the
 // map that's being read by the probes, indexed by the binary's inode number `ino`.
-func addInspectionResultToMap(offsetsDataMap *ebpf.Map, binID utils.PathIdentifier, result *bininspect.Result) error {
+func addInspectionResultToMap(offsetsDataMap, pidToDeviceInodeMap *ebpf.Map, filePath utils.FilePath, result *bininspect.Result) error {
 	offsetsData, err := inspectionResultToProbeData(result)
 	if err != nil {
 		return fmt.Errorf("error while parsing inspection result: %w", err)
 	}
 
 	key := &gotls.TlsBinaryId{
-		Id_major: unix.Major(binID.Dev),
-		Id_minor: unix.Minor(binID.Dev),
-		Ino:      binID.Inode,
+		Id_major: unix.Major(filePath.ID.Dev),
+		Id_minor: unix.Minor(filePath.ID.Dev),
+		Ino:      filePath.ID.Inode,
 	}
 	if err := offsetsDataMap.Put(unsafe.Pointer(key), unsafe.Pointer(&offsetsData)); err != nil {
-		return fmt.Errorf("could not write binary inspection result to map for binID %v: %w", binID, err)
+		return fmt.Errorf("could not write binary inspection result to map for binID %v (pid %v): %w", filePath.ID, filePath.PID, err)
 	}
 
+	if err := pidToDeviceInodeMap.Put(unsafe.Pointer(&filePath.PID), unsafe.Pointer(key)); err != nil {
+		return fmt.Errorf("could not write pid to device/inode (%s) map for pid %v: %w", filePath.ID.String(), filePath.PID, err)
+	}
 	return nil
 }
 
-func removeInspectionResultFromMap(offsetsDataMap *ebpf.Map, binID utils.PathIdentifier) {
+func removeInspectionResultFromMap(offsetsDataMap, pidToDeviceInodeMap *ebpf.Map, filePath utils.FilePath) {
 	key := &gotls.TlsBinaryId{
-		Id_major: unix.Major(binID.Dev),
-		Id_minor: unix.Minor(binID.Dev),
-		Ino:      binID.Inode,
+		Id_major: unix.Major(filePath.ID.Dev),
+		Id_minor: unix.Minor(filePath.ID.Dev),
+		Ino:      filePath.ID.Inode,
 	}
 	if err := offsetsDataMap.Delete(unsafe.Pointer(key)); err != nil {
-		log.Errorf("could not remove inspection result from map for ino %v: %s", binID, err)
+		log.Errorf("could not remove inspection result from map for ino %v: %s", filePath.ID, err)
+		return
 	}
+	_ = pidToDeviceInodeMap.Delete(unsafe.Pointer(&filePath.PID))
```

Comment:
In the paths where we don't provide a PID (cleanup case), this will be called with a zero PID, perhaps we should ignore that explicitly rather than attempting to delete it from the map?